ResearchKit/Common/ORKAnswerFormat.m

```diff
@@ -1887,39 +1908,40 @@ - (instancetype)copyWithZone:(NSZone *)zone {
 - (BOOL)isAnswerValid:(id)answer {
     BOOL isValid = NO;
     if ([answer isKindOfClass:[NSString class]]) {
-        return [self isAnswerValidWithString:(NSString *)answer];
+        isValid = [self isAnswerValidWithString:(NSString *)answer];
     }
     return isValid;
 }
 
 - (BOOL)isAnswerValidWithString:(NSString *)text {
-    if (text.length > 0) {
-        return ([self isTextLengthValidWithString:text] && [self isEmailAddressValidWithString:text]);
+    if (text) {
+        BOOL isValid = NO;
+        if (self.regex) {
+            NSString *regExPattern = self.regex;
+            NSRegularExpression *regEx = [[NSRegularExpression alloc] initWithPattern:regExPattern options:NSRegularExpressionCaseInsensitive error:nil];
+            NSUInteger regExMatches = [regEx numberOfMatchesInString:text options:0 range:NSMakeRange(0, [text length])];
+            isValid = (regExMatches != 0);
+        } else if (text.length > 0) {
+	        isValid = [self isTextLengthValidWithString:text];
+    	} else {
+            isValid = YES;
+        }
+        return isValid;
+    } else {
+        return YES;
     }
-    
-    return YES;
 }
 
 - (BOOL)isTextLengthValidWithString:(NSString *)text {
     return (_maximumLength == 0 || text.length <= _maximumLength);
 }
 
-- (BOOL)isEmailAddressValidWithString:(NSString *)text {
-    if (self.isEmailAddress) {
-        NSPredicate *emailValidationTest = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", EmailValidationRegex];
-        return [emailValidationTest evaluateWithObject:text];
-    }
-    
-    return YES;
-}
-
 - (NSString *)localizedInvalidValueStringWithAnswerString:(NSString *)text {
     NSString *string = nil;
     if (![self isTextLengthValidWithString:text]) {
         string = [NSString stringWithFormat:ORKLocalizedString(@"TEXT_ANSWER_EXCEEDING_MAX_LENGTH_ALERT_MESSAGE", nil), ORKLocalizedStringFromNumber(@(_maximumLength))];
```

Comment:
@umerkhan-apple The existing invalid e-mail alert is "Invalid e-mail address: your text here" which requires additional formatting to include the actual text the user entered.  In other locations we have static invalid messages like "Enter a valid URL."

Which style is preferable?  Should we:
1) Change the e-mail address INVALID_EMAIL_ALERT_MESSAGE to be a static alert like "Invalid e-mail address"
2) Change this code to always assume dynamic alerts and update existing alerts to match the existing INVALID_EMAIL_ALERT_MESSAGE e.g. "Invalid URL: your text here"
```
string = [NSString stringWithFormat:ORKLocalizedString(_invalidMessage, nil), text];
```
3) Try to detect which kind of invalid message has been specified by looking if _invalidMessage is already a localized string.